File: python/0104 MaximumDepthofBinaryTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        
        queue = [(root, 1)]
        height = 0
        
        while queue:
            node, depth = queue.pop(0)
            height = max(depth,height)
            
            if node.left:
                queue.append((node.left, depth+1))
            if node.right:
                queue.append((node.right, depth+1))
                
        return height

File: javascript/0104 Maximum Depth of Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
    let best = 0;

    if(root){
        let queue = [Object.freeze([root,1])];

        while(queue.length>0){

            let node = queue.shift();

            let current = node[0];
            best = node[1];

            if(current.left) queue.push(Object.freeze([current.left,best+1]))
            if(current.right) queue.push(Object.freeze([current.right,best+1]))

        }
    }
    return best;
};
